src/ReplayTimeline.cc

```diff
@@ -1668,4 +1716,50 @@ ReplayTimeline::Mark ReplayTimeline::set_short_checkpoint() {
   return reverse_exec_short_checkpoint;
 }
 
+ReplayTimeline::ProtoMark ReplayTimeline::ProtoMark::from_serialized_markdata(const MarkData& md) {
+  auto proto_mark = ProtoMark{ MarkKey{ md.time, md.ticks, ReplayStepKey{ (ReplayTraceStepType)md.step_key } } };
+  proto_mark.regs = md.regs;
+  proto_mark.return_addresses = md.return_addresses;
+  return proto_mark;
+}
+
+std::shared_ptr<ReplayTimeline::Mark> ReplayTimeline::find_closest_mark_with_clone(const Mark &mark) {
+  if(marks_with_checkpoints.empty()) return nullptr;
+
+  const MarkKey* k = &marks_with_checkpoints.begin()->first;
+  for(const auto& kvp : marks_with_checkpoints) {
+    if(kvp.first < mark.ptr->proto.key && kvp.first > *k) {
+      k = &kvp.first;
+    }
+  }
+
+  auto marks_found = std::find_if(std::cbegin(marks), std::cend(marks), [&](auto& kvp){
+    return kvp.first == *k;
+  });
+
+  if(marks_found == std::end(marks)) return nullptr;
+
+  for(auto it = std::rbegin(marks_found->second); it != std::rend(marks_found->second); it++) {
+    DEBUG_ASSERT(it->get() != nullptr);
+    if((*it)->checkpoint && (*it)->checkpoint->partially_initialized()) {
```

Comment:
In retrospect it solves nothing, I had gated it behind that, to make sure it's not fully initialized, but since it's a checkpoint, by definition it won't be, so there's no reason to actually use this here.